components/modules/touch.c

```diff
@@ -0,0 +1,745 @@
+
+        } else {
+          // don't push false values for now to reduce memory usage
+          //   l_pushtableintkeybool(L, i, false);
+          //   if (touch_self->is_debug) ESP_LOGI(TAG, "Pushed key %d, bool %d", i, false);
+        }
+      }
+
+      // call the cb_ref with one argument
+      /* do the call (1 argument, 0 results) */
+      if (lua_pcall(L, 1, 0, 0) != 0) {
+        ESP_LOGI(TAG, "error running callback function `f': %s", funcName);
+      }
+    }
+  }
+
+  //} // old for loop
+
+}
+
+
+/* 
+Lua sample code:
+tp = touch.create({
+  pad = {8,9}, -- pad = 0 || {0,1,2,3,4,5,6,7,8,9} 0=GPIO4, 1=GPIO0, 2=GPIO2, 3=GPIO15, 4=GPIO13, 5=GPIO12, 6=GPIO14, 7=GPIO27, 8=GPIO33, 9=GPIO32
+  cb = onTouch, -- Callback on touch event. Optional. If you do not provid you can just call touch:read(). If you do add a callback then TOUCH_FSM_MODE_TIMER is used.
+  intrInitAtStart = false, -- Turn on interrupt at start. Default to true. Set to false in case you want to config first. Turn it on later with tp:intrEnable()
+  thres = {[8]=300, [9]=500}, -- thres = 400 || {[8]=300, [9]=500} Pass in one uint16 (max 65536) or pass in table of pads with threshold value for each pad. Threshold of touchpad count. To get the right number, watch your touch pads for the high/low value. Run touch.test() to analyze all pads for their values.
+  thresTrigger = TOUCH_TRIGGER_BELOW, -- Touch interrupt will happen if counter value is below or above threshold. TOUCH_TRIGGER_BELOW or TOUCH_TRIGGER_ABOVE 
+  slope = 4, -- Touch sensor charge / discharge speed, 0-7 where 0 is no slope (so counter will always be 0), 1 is slow (lower counter), 7 is fastest (higher counter). Run the touch.test(pad) to test all combinations to pick best one.
+  lvolt = TOUCH_LVOLT_0V5, -- Touch sensor low reference voltage TOUCH_LVOLT_0V4, TOUCH_LVOLT_0V5, TOUCH_LVOLT_0V6, TOUCH_LVOLT_0V7 
+  hvolt = TOUCH_HVOLT_2V7, -- Touch sensor high reference voltage TOUCH_HVOLT_2V4, TOUCH_HVOLT_2V5, TOUCH_HVOLT_2V6, TOUCH_HVOLT_2V7
+  atten = TOUCH_HVOLT_ATTEN_1V, -- Touch sensor high reference voltage attenuation TOUCH_HVOLT_ATTEN_0V, TOUCH_HVOLT_ATTEN_0V5, TOUCH_HVOLT_ATTEN_1V, TOUCH_HVOLT_ATTEN_1V5 
+  isDebug = true
+})
+*/
+static int touch_create( lua_State *L ) {
+
+  // Check if we are out of objects. There are only 10 pads, so should never get this, but check anyhow.
+  if (touch_self != NULL) {
+    luaL_error(L, "Only 1 touch pad object allowed in touch library.");
+  }
+
+  // Presume we'll get a good create, so go ahead and make our touch_t object 
+  touch_struct_t tpObj = {.cb_ref=LUA_NOREF, .is_initted=false};
+  touch_t tp = &tpObj;
+
+  // Now go through our table that was passed in with our parameters
+  luaL_checkstack(L, 5, "out of mem");
+  int i = 1;
+
+  // checks to make sure we got a table
+  luaL_checkanytable(L, i);
+
+  // get the field isDebug so we know whether to show config info or not during init
+  tp->is_debug = false;
+  lua_getfield(L, i, "isDebug");
```

Comment:
I modeled the code after the gpio and tmr modules, so I didn't know the opt_* methods existed. I can go check that out.

Reply:
Ok, refactored the create() code to use the opt_* methods. Should be good now. Couldn't use it for all elements like pad where it can be an int or a table, but it worked for majority.